Using a kernel option and userland build option to make save/restore optional for now.

bsdjhb opened the pull request.

bsdjhb committed:
Some style fixes in libvmmapi.

- Move debug printfs under #ifdef SNAPSHOT_DEBUG
- Fail with ENOSPC if the memory snapshot file is too small.

bsdjhb committed:
Add missing break.

This one happens to be hamless.

bsdjhb committed:
Add a kmod.opts.mk.

This Makefile sets KERN_OPTS.  This permits kernel module Makefiles to
use KERN_OPTS to control the value of variables such as SRCS that are
used by bsd.kmod.mk for KERN_OPTS values that honor WITH/WITHOUT
options for standalone builds.

bsdjhb committed:
Wrap kernel support for save/restore under a BHYVE_SNAPSHOT kernel option.

bsdjhb commented:
This is not yet ready to merge, but the kernel bits could be tested perhaps.  Also, if you have comments or questions on the changes, please let me know.

darius-m commented on lib/libvmmapi/vmmapi.c:
```diff
@@ -461,27 +462,26 @@ vm_rev_map_gpa(struct vmctx *ctx, void *addr)
-		return (EINVAL);
 
-	strlcpy(buf, ctx->name, max_len);
+	if (strlcpy(buf, ctx->name, max_len) >= max_len)
```
This has the potential of overwriting data beyond the buffer's limit before checking bounds.

bsdjhb replied:
strlcpy won't write more than max_len bytes (and it ensure the result is null-terminated), but it returns the length it would have written that can be used to detect overflow.  The original code could have safely used strcpy since it did the length check ahead of time, but since it is already using strlcpy() it can make use of the return value to avoid walking the string more than once.

darius-m commented on lib/libvmmapi/vmmapi.c:
```diff
@@ -1548,87 +1548,96 @@ vm_restart_instruction(void *arg, int vcpu)
 
-	if ( lseek(fd, file_offset , SEEK_SET) < 0) {
+	if (lseek(fd, file_offset, SEEK_SET) == -1) {
+#ifef SNAPSHOT_DEBUG
```
Misspelling of "#ifdef"

bsdjhb replied:
Fixed.

darius-m commented on lib/libvmmapi/vmmapi.c:
```diff
@@ -1548,87 +1548,96 @@ vm_restart_instruction(void *arg, int vcpu)
 
-	if (vm_mem_read_from_file(vmmem_fd, ctx->baseaddr,
-				0, ctx->lowmem) != 0) {
+	if (vm_mem_read_from_file(vmmem_fd, ctx->baseaddr, 0, ctx->lowmem) !=
```
This may fit on one line by using the implied (`val != 0`) of the `if` conditional

bsdjhb replied:
The normal style in FreeBSD is to only use boolean expressions in if () and while ().  If vm_mem_read_from_file() returned a bool then using it directly would work, but if it returns an errno value comparing against 0 explicitly is the normal style.  I suppose we could make it return bool in this case since it is used internally and only returns either 0 or -1.

darius-m commented on lib/libvmmapi/vmmapi.c:
```diff
@@ -1548,87 +1548,96 @@ vm_restart_instruction(void *arg, int vcpu)
-	error = ioctl(ctx->fd, VM_RESTORE_TIME, &dummy);
-
-	return (error);
+	return (ioctl(ctx->fd, VM_RESTORE_TIME, &dummy));
```
The dummy argument is not used. We can give it up completely, so the function will only contain `return (ioctl(ctx->fd, VM_RESTORE_TIME));`

bsdjhb replied:
We will have to edit the VM_RESTORE_TIME definition to not take an 'int' first, but that change is a good cleanup change to do.

bsdjhb committed:
Add a new src.conf option BHYVE_SNAPSHOT.

This currently is disabled by default, but it enables support
for snapshots in bhyve and bhyvectl.

bsdjhb committed:
Make save/restore code in bhyve(8) conditional on WITH_BHYVE_SNAPSHOT.

bsdjhb committed:
Make bhyvectl's save/restore support conditional on WITH_BHYVE_SNAPSHOT.

darius-m commented on usr.sbin/bhyve/pci_emul.c:
```diff
@@ -1937,8 +1937,9 @@ INOUT_PORT(pci_cfgdata, CONF1_DATA_PORT+1, IOPORT_F_INOUT, pci_emul_cfgdata);
+#ifdef BHYVE_SNAPSHOT
 /*
- * Saves/restores PCI device emulated state. Returns 0 on success.
+ * Saves/restores PCI device emulated state.  Returns 0 on success.
```
Why the double blank?

bsdjhb replied:
I can revert this bit.

Some background: some folks use a double-space after full stops to indicate it is a "bigger" gap between sentences than between words when using a monospace font.  I tend to prefer that style, but it isn't a hard rule.  I looked at some other comments in this file and they all use a single space, so being consistent with the rest of the file is more important.

bsdjhb committed:
Fixups for libvmmapi change.

bsdjhb committed:
Fixup to revert whitespace change in pci_emul.c.

darius-m commented on usr.sbin/bhyve/bhyverun.c:
```diff
@@ -60,6 +69,12 @@ __FBSDID("$FreeBSD$");
 #include <sysexits.h>
 #include <stdbool.h>
 #include <stdint.h>
+#iifdef BHYVE_SNAPSHOT
```
Typo

bsdjhb replied:
Fixed

darius-m commented on lib/libvmmapi/vmmapi.c:
```diff
@@ -1559,19 +1559,19 @@ vm_snapshot_req(struct vm_snapshot_meta *meta)
 }
 
-static int
+static bool
```
`bool`, `true`, and `false` are not defined. Missing `#include <stdbool.h>`

bsdjhb replied:
Fixed

bsdjhb committed:
Another fixup for libvmmapi.

bsdjhb committed:
A fixup for bhyve.

bsdjhb commented:
Once it compiles and is tested I'll use rebase to merge the fixups into their respective commits.

bsdjhb committed:
Document --checkpoint and --suspend in bhyvectl(8).

aleksandr-fedorov commented on usr.sbin/bhyve/block_if.c:
```diff
@@ -935,17 +929,19 @@ blockif_snapshot_req(struct blockif_req *br, struct vm_snapshot_meta *meta)
+
+		/* We assume the iov is a guest-mapped address. */
 		SNAPSHOT_GUEST2HOST_ADDR_OR_LEAVE(iov->iov_base, iov->iov_len,
 			false, meta, ret, done);
```
It seems, that this "false" also requires to include <stdbool.h>

aleksandr-fedorov replied:
And in other places.

aleksandr-fedorov replied:
Can we change 'restore_null' type to bool and include <stdbool.h> in sys/amd64/include/vmm_snapshot.h?

`
int vm_snapshot_guest2host_addr(void **addrp, size_t len, int restore_null,
                 struct vm_snapshot_meta *meta);
`

bsdjhb replied:
I think this is fallout from removing the #include <ucl.h> from bhyverun.h.  I've added #includes of <stdbool.h>

bsdjhb replied:
Hmm, I hadn't seen the more recent  about the prototype, let me look at that.

bsdjhb replied:
Interestingly, the in-kernel version of this isn't used, it's only used in userland in bhyve.

darius-m replied:
The kernel does not keep mappings of guest memory, only emulated devices that reside in the userspace do. The function was created to get guest memory addresses,, so it is not needed.

bsdjhb committed:
Fixups to bhyve(8) build.

The stdbool.h are due to removing the nested include of ucl.h from
bhyverun.h I think.

bsdjhb committed:
Document -r in bhyve(8).

bsdjhb committed:
Don't provide a kernel stub for vm_snapshot_guest2host_addr.

This ensures any uses of it in the kernel will fail with a link error rather
than failing silently.

bsdjhb committed:
Change the restore_null argument to vm_snapshot_guest2host_addr to bool.

This is what all the callers and the implementation expect.

darius-m commented:
Tested with a FreeBSD and a Windows guest, and works as expected.
